### PR TITLE
Switch to codecov

### PR DESCRIPTION
**Pull request by dmoody256**
This PR switches to codecov which allows for coverage reports submitted from both Appveyor and Travis so we can get a complete coverage including windows code and tests.

* Fixes some minor issues with the files omitted for coverage calculation
* disables coveralls and switches to using codecov for coverage reporting
* adds py 2.7 and 3.6 on the Visual Studio 2015 image, and maintains py 2.7, 3.5, 3.6, 3.7 on the 2017 image
* adds additional coverage for py 3.6 on Travis, and compresses coverage builds into 4 jobs each for speed
* calculates coverage even if a single test fails, as this can still be useful information
* adds the codecov.yml for source level coverage configuration.
* Adds appveyor Ubuntu configuration as an optional environment, but disabled for now

Codecov example report:
https://codecov.io/gh/SCons/scons

### CODE REVIEW

**Comment by bdbaddog** [comment]
Why do linux on both appveyor and travis?

**Comment by bdbaddog** [comment]
I'd say leave the logic in for appveyor, but comment it out for now? Sounds like this would hold up any other PR's builds until extra runtime on appveyor which is already a the slowest part of the build..

**Comment by dmoody256** [comment]
regarding the failures in the Visual Studio 2015 appveyor image:

- the mingw failure is related to the PR https://github.com/SCons/scons/pull/3270
- the java failures are due to the test framework selecting a different javac (1.7) and java (1.6) from the where_java calls. I wrote up an issue here: https://github.com/SCons/scons/issues/3276

File: .appveyor.yml [hunk 1]
```diff
@@ -1,83 +1,248 @@
...
-  - set STATIC_DEPS=true & pip install lxml
-  - choco install --allow-empty-checksums dmd
-  - choco install --allow-empty-checksums ldc
-  - choco install --allow-empty-checksums swig
-  - choco install --allow-empty-checksums vswhere
-  - choco install --allow-empty-checksums xsltproc
+  ### WINDOWS ###
+  # add python and python user-base to path for pip installs
+  - cmd: "C:\\%WINPYTHON%\\python.exe --version"
+  - cmd: for /F "tokens=*" %%g in ('C:\\%WINPYTHON%\\python.exe -m site --user-base') do (set PYUSERBASESITE=%%g)
```
Comment (bdbaddog) [inline .appveyor.yml]: Why use user-base?

Comment (dmoody256) [reply .appveyor.yml]: to avoid any permissions issues

**Comment by bdbaddog** [comment]
Check the appveyor run? some java test failed looking for java version 51?

**Comment by dmoody256** [comment]
> Check the appveyor run? some java test failed looking for java version 51?

That's the java issue I mentioned earlier: #3276. I was looking into it and the binaries are all present on that appveyor image and working, so I think there is something going on with the test framework, haven't nailed it down yet though.

**Comment by bdbaddog** [comment]
Is this supposed to only test with py27 and 36, and not 3.5 or 3.7?

**Comment by dmoody256** [comment]
> Is this supposed to only test with py27 and 36, and not 3.5 or 3.7?

I removed 3.5 and 3.7 because the time it takes to run the CI with one job. Those are tested on Linux in Travis, so I am not sure how much more value they are adding on windows. With this PR as it stands, Appveyor takes 2 hours, instead of the 4 currently. I can add them back if you think its worth it.

**Comment by bdbaddog** [comment]
Since the PR title is switch to codecov, it's be best if this PR only did that and not change anything else.
Smaller PRs that do one thing are the preferred.
So for this PR please change it back to running the same Python versions as before.

**Comment by bdbaddog** [comment]
looks like win32/mingw.py test is failing on appveyor? can you take a look

**Comment by dmoody256** [comment]
> looks like win32/mingw.py test is failing on appveyor? can you take a look

Thats from the issue in PR: #3270

Also the VS 2015 exec test failed related to license expired, I notified Appveyor:
https://help.appveyor.com/discussions/problems/19717-visual-studio-2015-trial-license-has-expired

**Comment by dmoody256** [comment]
> Since the PR title is switch to codecov, it's be best if this PR only did that and not change anything else.
> Smaller PRs that do one thing are the preferred.
> So for this PR please change it back to running the same Python versions as before.

OK just to make sure I understand, the CI should test py 2.7, 3.5, 3.6, 3.7 on the Visual Studio 2017 image (the current config) and also test py 2.7, 3.6 on Visual Studio 2015 image?

**Comment by bdbaddog** [comment]
Can you update the summary for this PR to indicate all the included changes?
Likely we should update CHANGES.txt since the CI tasks being run is also(?) being changed?

**Comment by dmoody256** [comment]
> Can you update the summary for this PR to indicate all the included changes?

sure

> Likely we should update CHANGES.txt since the CI tasks being run is also(?) being changed?

I thought CHANGES.txt was related to src tree changes, and since this is CI is not documented?

**Comment by bdbaddog** [comment]
In this case since it's a change in the environments we regularly test in, probably worth noting.

**Comment by bdbaddog** [comment]
any chance you can pop on IRC?

**Comment by bdbaddog** [comment]
Per IRC discussion only expected tests failed. I'm going to merge this and then we'll address the failures with other PRs.
Thanks for the good work!

**Pull request merged**